src/librustc/traits/select.rs

```diff
@@ -78,8 +77,40 @@ pub struct SelectionContext<'cx, 'tcx:'cx> {
     /// there is no type that the user could *actually name* that
     /// would satisfy it. This avoids crippling inference, basically.
     intercrate: bool,
+
+    /// Obligations generated during inference. If None, no obligations are to be collected (e.g.
+    /// this is swapped out to a `None` when probing).
+    inferred_obligations: Option<PredicateObligations<'tcx>>
```

Comment:
The last time I remember digging into this PR and trying to pull this field out and make the flow of obligations explicit, and it wound up leading me to problems in project.rs (the fix for this *still* hasn't landed; grr). So it might even be a bit hard to do this PR without those fixes having landed, or else we'd need to add more `assert-is-empty` points in `project.rs`.